counsel.el

```diff
@@ -3294,20 +3294,18 @@ attachment directory associated with the current buffer, all
 contained files are listed, so the return value could conceivably
 include attachments of other Org buffers."
   (require 'org-attach)
-  (let* ((ids (let (res)
-                (save-excursion
-                  (goto-char (point-min))
-                  (while (re-search-forward "^:ID:[\t ]+\\(.*\\)$" nil t)
-                    (push (match-string-no-properties 1) res))
-                  (nreverse res))))
-         (files
-          (cl-remove-if-not
-           #'file-exists-p
-           (mapcar (lambda (id)
-                     (expand-file-name
-                      (concat (substring id 0 2) "/" (substring id 2))
-                      org-attach-directory))
-                   ids))))
+  (let ((files (let (res dir)
+                 (save-excursion
+                   (goto-char (point-min))
+                   (while (re-search-forward "^:\\(ATTACH_DIR\\|ID\\):[\t ]+\\(.*\\)$" nil t)
+                     (setq dir (match-string-no-properties 2))
+                     (when (string= "ID" (match-string-no-properties 1))
+                       (setq dir (expand-file-name
+                                  (concat (substring dir 0 2) "/" (substring dir 2))
+                                  org-attach-directory)))
```

Comment:
Why not use the `org-attach-dir` API instead of duplicating a subset of it?

Reply:
Not sure what's the best way to go. On the one hand, using `org-attach-dir` would simplify the code. On the other hand, it would unnecessarily parse the whole property block of the current entry when we have already matched the relevant property line.

Reply:
> Not sure what's the best way to go.

IME, Org is a complicated and fickle beast, so whenever there is a stable API, we should try to reuse it.

> On the one hand, using `org-attach-dir` would simplify the code.

And, more importantly, make it more correct and robust.

> On the other hand, it would unnecessarily parse the whole property block of the current entry when we have already matched the relevant property line.

Does this really cause a noticeable slowdown? I doubt it, but I could be wrong.

Reply:
I don't have a very large org file to test. Since `counsel-org-files` was not using `org-attach-dir` before, I would rather know what @abo-abo thinks before changing that.

Reply:
I'm fine with the current implementation. Org will not break this API unless they want to break the users' existing attachment directories.

Reply:
@abo-abo 

> Org will not break this API unless they want to break the users' existing attachment directories.

I never said Org might break the `ATTACH_DIR` API. What I said is:

> Why not use the `org-attach-dir` API instead of duplicating a subset of it?

Emphasis on "subset". The function `org-attach-dir` handles not only `ATTACH_DIR`, but also `ATTACH_DIR_INHERIT` and related Org settings. If it is possible to reuse the longstanding `org-attach-dir` API instead of emulating a subset of it, then we absolutely should.